Prepared queries without variables

msolomon opened the pull request:
This is a quick-and-dirty sketch for what #277 might look like. 

The basic approach is to add a new `Executor.prepare` method that doesn't accept variables, and add `preparedQueryReducers` to `Executor`.

Query reducers are run depending on the scenario:

* If we are preparing a query *without* known variables
  + Run `preparedQueryReducers` now and defer `queryReducers` to runtime
* If we are preparing a query *with* known variables
  + Run `preparedQueryReducers` and `queryReducers` now
* If we are executing an unprepared query:
  + Run `queryReducers`

The implementation is complete enough for my use case, but is not merge-ready.

What do you think about separating preparation-time from run-time like this?

OlegIlyenko commented:
Thanks a lot for working on it!

In general, I like the approach. There is one concern I have with making this feature a part of `Executor.prepare`. I feel that in a long term it can be error-prone and a bit misleading. This breaks the contract of `Args` - in terms of type-safety this means that things that were non-optional suddenly will throw exceptions if used in particular way (in complexity function). I do think that the feature itself is quite valuable and this might be a suitable tradeoff. But I would like to make it a bit more obvious when query reducers are executed without variables. I also would like to decouple this feature from the `prepare`, i think they have different semantics and meant to be used for different use-cases.

What do you think about making small refactoring and extracting `reduceQuerySafe` away from the `Executor`? It is quite independent from everything else, so I think it can be easily extracted elsewhere (like a companion object of `QueryReducer` or companion object of  `Executor` or somewhere else). After this is done, we can still use it from `Executor`, as before. But then we can also provide a nice API to execute query reducers against query document and schema with or without variables. This will make it quite explicit and will not couple this functionality to execution in any way.

On a side note. Using `Args.empty` in case of empty variables can be quite restrictive. We can still discover arguments that are defined in the query itself. I think this will require additional refactoring and might be considered an improvement of this feature. So we can tackle it as a separate step in different PR.

msolomon commented on src/main/scala/sangria/execution/QueryReducer.scala:
```diff
@@ -38,6 +37,8 @@ trait QueryReducer[-Ctx, +Out] {
 }
 
 object QueryReducer {
+  type ArgumentValuesFn = (ExecutionPath, List[Argument[_]], Vector[ast.Argument]) => Try[Args]
```
I found this helpful when reading the code

msolomon commented on src/main/scala/sangria/execution/QueryReducerExecutor.scala:
```diff
@@ -0,0 +1,179 @@
+
+object QueryReducerExecutor {
+
+  def reduceQueryWithoutVariables[Ctx, Root](
```
used when variables aren't known

msolomon commented on src/main/scala/sangria/execution/QueryReducerExecutor.scala:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  /** Returns either new Ctx or future of it (with time measurement) */
+  def reduceQuery[Ctx, Root, Val](
```
used when variables are known (prepare, execute)

msolomon commented:
Updated with your suggested model.

Some notes:

* Literal arguments defined in the query will now be discovered
* Added `QueryReducerExecutor` which holds static methods to run `QueryReducer`s over `Document`s and produce a user context
* `reduceQuerySafe` moved to `QueryReducerExecutor` and renamed to `reduceQuery`
* `reduceQuery` moved to `QueryReducerExecutor` and renamed to `reduceQueryUnsafe`

I also went through and added notes inline. I'd love to add inline Scaladocs, but for much of the existing code I don't feel qualified to describe the purpose and scope of each class.

I look forward to your comments!

OlegIlyenko commented on src/main/scala/sangria/schema/package.scala:
```diff
@@ -190,13 +190,13 @@ package object schema {
     arguments = IfArg :: Nil,
     locations = Set(DirectiveLocation.Field, DirectiveLocation.FragmentSpread, DirectiveLocation.InlineFragment),
-    shouldInclude = ctx ⇒ ctx.arg(IfArg))
+    shouldInclude = ctx ⇒ ctx.args.argOpt(IfArg).getOrElse(true))
```
yeah... I don't really like this, but I also don't see a better way at the moment. Maybe you can add a comment above to clarify why this code is like this and under which circumstances the argument would be empty. (the same below)

OlegIlyenko commented on src/main/scala/sangria/execution/DeprecationTracker.scala:
```diff
@@ -9,18 +9,18 @@ trait DeprecationTracker {
 }
 
-object PrintingDeprecationTracker extends DeprecationTracker {
+class PrintingDeprecationTracker(printline: String => Unit) extends DeprecationTracker {
```
👍 Since it is now changed, maybe we can rename it a bit. I would suggest:

```scala
class LoggingDeprecationTracker(logFn: String ⇒ Unit)
```

OlegIlyenko commented on src/main/scala/sangria/execution/DeprecationTracker.scala:
```diff
@@ -9,18 +9,18 @@ trait DeprecationTracker {
 object DeprecationTracker {
   val empty = NilDeprecationTracker
-  val print = PrintingDeprecationTracker
+  val print = new PrintingDeprecationTracker(println)
```
👍

OlegIlyenko commented on src/main/scala/sangria/execution/QueryReducerExecutor.scala:
```diff
@@ -0,0 +1,179 @@
+    schema: Schema[Ctx, Root],
+    queryAst: ast.Document,
+    userContext: Ctx,
+    root: Root,
```
You don't really need the root value, isn't it? (it's not used anywhere)

OlegIlyenko commented on src/test/scala/sangria/execution/QueryReducerSpec.scala:
```diff
@@ -789,4 +790,71 @@ class QueryReducerSpec extends WordSpec with Matchers with FutureResultSupport {
+          .await
+        assert(res._1.nums == Seq(100, 1, 1, 1, 1), "expected 4 nodes to be visited")
+      }
+    }
```
Maybe you can also add extra tests with some existing query reduces and without variables? (`MeasureComplexity` in particular since it would be the main use-case, as far as I understand) You probably will need to define different schema just for this test because complexity function would have a different implementation

msolomon replied:
I added a test with MeasureComplexity, let me know if you had more in mind

OlegIlyenko commented on src/main/scala/sangria/execution/QueryReducerExecutor.scala:
```diff
@@ -0,0 +1,179 @@
+        .recover { case error: Throwable ⇒ throw QueryReducingError(error, exceptionHandler) }
+    } else Future.successful(userContext → TimeMeasurement.empty)
+
+  private def reduceQueryUnsafe[Ctx, Val](
```
This probably was not the easiest part of the codebase to read through :) It is quite mutable and untyped. Not very proud of this code, but this approach helps to reduce the object allocation. The main idea was to reduce performance impact as much as possible.

msolomon commented on src/main/scala/sangria/execution/Executor.scala:
```diff
@@ -90,18 +84,20 @@ case class Executor[Ctx, Root](
-            fieldCollector, marshaller, unmarshalledVariables, tpe, fields, c._1, root, scheme, validationTiming, c._2))
-        case (ctx: Ctx @unchecked, timing: TimeMeasurement) ⇒
+      } yield {
+        val argumentValuesFn: QueryReducer.ArgumentValuesFn =
```
Ah, yes. I used to inject this function, but later changed my mind

msolomon commented:
Any thoughts on the latest updates?

msolomon commented:
@OlegIlyenko is there anything I can do to make this easier to review?

OlegIlyenko commented:
Hey Mike, sorry for the delay. I actually wanted to come back to it during the last week, but haven't managed it 😅 I already went quickly through it and it looks great! 👍 Just wanted to take some time give it a final thorough look before merging. I will try to do it during the next few days and then merge it.

OlegIlyenko commented on src/main/scala/sangria/schema/package.scala:
```diff
@@ -190,13 +190,19 @@ package object schema {
-    shouldInclude = ctx ⇒ ctx.arg(IfArg))
+    // if we don't know if we should include it, then we should include it:
+    // ValueCollector will fail before we get here if values must be known, such as when preparing or executing a query,
+    // but for e.g. running a QueryReducer without known variables, we must be conservative
```
👍 Thanks
